### DISCUSSION TITLE
"Released too many references to asset!" or "Unimplemented asset type!"

### DISCUSSION THREAD

**Discussion by smasonuk** [StereoKit/StereoKit#463] [Q&A]
I've noticed that often when I shut down the stereokit app on the desktop I get this error 'Released too many references to asset!' or ' Unimplemented asset type!' in the output. Anyone know if this is something I'm doing wrong, and how to track down where the issue is. 

This is the full output:
[SK diagnostic] Initializing StereoKit v0.3.6 Win32 x64...
[SK diagnostic] Initializing Assets
[SK diagnostic] Initializing Platform
Error [GENERAL |  | OpenXR-Loader] : RuntimeManifestFile::FindManifestFiles - failed to find active runtime file in registry
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : RuntimeInterface::LoadRuntimes - unknown error
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : RuntimeInterface::LoadRuntimes - failed to load a runtime
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : Failed loading runtime information
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : xrCreateInstance failed
[SK diagnostic] sk_gpu: Using Direct3D 11: Intel(R) UHD Graphics 620
[SK diagnostic] Starting mixed reality mode
Error [GENERAL |  | OpenXR-Loader] : RuntimeManifestFile::FindManifestFiles - failed to find active runtime file in registry
Error [GENERAL | xrEnumerateInstanceExtensionProperties | OpenXR-Loader] : RuntimeInterface::LoadRuntimes - unknown error
Error [GENERAL | xrEnumerateInstanceExtensionProperties | OpenXR-Loader] : RuntimeInterface::LoadRuntimes - failed to load a runtime
Error [GENERAL | xrEnumerateInstanceExtensionProperties | OpenXR-Loader] : Failed to find default runtime with RuntimeInterface::LoadRuntime()
Error [GENERAL | xrEnumerateInstanceExtensionProperties | OpenXR-Loader] : Failed querying extension properties
Error [GENERAL |  | OpenXR-Loader] : RuntimeManifestFile::FindManifestFiles - failed to find active runtime file in registry
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : RuntimeInterface::LoadRuntimes - unknown error
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : RuntimeInterface::LoadRuntimes - failed to load a runtime
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : Failed loading runtime information
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : xrCreateInstance failed
[SK info] Couldn't create OpenXR instance [XR_ERROR_RUNTIME_UNAVAILABLE], is OpenXR installed and set as the active runtime?
[SK info] MixedReality display mode failed, falling back to Flatscreen
[SK diagnostic] Starting flatscreen mode
[SK diagnostic] Created swapchain: 1280x701 color:rgba32_linear depth:depth32
[SK diagnostic] Initializing Audio
[SK diagnostic] ISAC audio backend not available, falling back to miniaudio! It's likely the device doesn't have Windows Sonic enabled, which can be found under Settings->Sound->Device Properties->Spatial Sound.
[SK info] Using audio backend: WASAPI
[SK diagnostic] Initializing Defaults
[SK diagnostic] Initializing World
[SK diagnostic] Initializing Sprites
[SK diagnostic] Initializing Lines
[SK diagnostic] Initializing UI
[SK diagnostic] Initializing Renderer
[SK diagnostic] Initializing Physics
[SK diagnostic] Initializing Input
[SK info] Initialization successful
[SK error] Released too many references to asset!

**Comment by maluoi** [comment]
Does this also happen for you in the v0.3.7 preview builds? I've done some work in this area to correct some issues that async textures recently may have introduced.

If it's still happening in v0.3.7-preview.3, I might want to know a bit more about what you use pattern is! :)

**Reply by smasonuk** [reply]
Yeah I've tried preview 2 and preview 3. I can't reproduce it every time. I'll see if I can track down roughly where it's happening and put together a small demo app.. 

I just tried again on preview 3 and I got a new error.. "Destroying asset '(null)' that still has references"

[SK diagnostic] Initializing StereoKit v0.3.7-preview.3 Win32 x64...
[SK diagnostic] Initializing Platform
Error [GENERAL |  | OpenXR-Loader] : RuntimeManifestFile::FindManifestFiles - failed to find active runtime file in registry
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : RuntimeInterface::LoadRuntimes - unknown error
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : RuntimeInterface::LoadRuntimes - failed to load a runtime
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : Failed loading runtime information
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : xrCreateInstance failed
[SK diagnostic] sk_gpu: Using Direct3D 11: Intel(R) UHD Graphics 620
[SK diagnostic] Starting mixed reality mode
Error [GENERAL |  | OpenXR-Loader] : RuntimeManifestFile::FindManifestFiles - failed to find active runtime file in registry
Error [GENERAL | xrEnumerateInstanceExtensionProperties | OpenXR-Loader] : RuntimeInterface::LoadRuntimes - unknown error
Error [GENERAL | xrEnumerateInstanceExtensionProperties | OpenXR-Loader] : RuntimeInterface::LoadRuntimes - failed to load a runtime
Error [GENERAL | xrEnumerateInstanceExtensionProperties | OpenXR-Loader] : Failed to find default runtime with RuntimeInterface::LoadRuntime()
Error [GENERAL | xrEnumerateInstanceExtensionProperties | OpenXR-Loader] : Failed querying extension properties
Error [GENERAL |  | OpenXR-Loader] : RuntimeManifestFile::FindManifestFiles - failed to find active runtime file in registry
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : RuntimeInterface::LoadRuntimes - unknown error
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : RuntimeInterface::LoadRuntimes - failed to load a runtime
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : Failed loading runtime information
Error [GENERAL | xrCreateInstance | OpenXR-Loader] : xrCreateInstance failed
[SK info] Couldn't create OpenXR instance [XR_ERROR_RUNTIME_UNAVAILABLE], is OpenXR installed and set as the active runtime?
[SK info] MixedReality display mode failed, falling back to Flatscreen
[SK diagnostic] Starting flatscreen mode
[SK diagnostic] Created swapchain: 1280x701 color:rgba32_linear depth:depth32
[SK diagnostic] Initializing Audio
[SK diagnostic] ISAC audio backend not available, falling back to miniaudio! It's likely the device doesn't have Windows Sonic enabled, which can be found under Settings->Sound->Device Properties->Spatial Sound.
[SK info] Using audio backend: WASAPI
[SK diagnostic] Initializing Assets
[SK diagnostic] Initializing Defaults
[SK diagnostic] Initializing World
[SK diagnostic] Initializing Sprites
[SK diagnostic] Initializing Lines
[SK diagnostic] Initializing UI
[SK diagnostic] Initializing Renderer
[SK diagnostic] Initializing Physics
[SK diagnostic] Initializing Input
[SK info] Initialization successful
[SK error] Released too many references to asset!
[SK error] Destroying asset '(null)' that still has references!

**Reply by maluoi** [reply]
Okay, can you provide a bit of detail about what your app is doing? In particular, what sort of assets are you working with, and how are you working with them. A sample that repros it occasionally will probably be enough too.

Since this is likely threading related, I've generally found this type of issue to be tricky to isolate and reproduce consistently! I do see that you're on an Intel integrated chip, so I'm guessing this is a bit of a slower of a computer than what I normally dev with. I was observing in v0.3.6 that issues did happen a lot less with faster computers. I thought I had eliminated the issue, but perhaps I just made it more rare?

Either way, thanks for bringing this to my attention! It's definitely not you, I hadn't seen it in ages and I thought it was gone :(

**Reply by smasonuk** [reply]
I think I've narrowed it down to the part of the application which displays molecules. This has been working ok up until today as far as I've noticed.. I'll try and exact the code and create a simple app which can reproduce the issue..

The molecule browser is mostly the code from here 'https://github.com/cefoot/Molecule-Browser' but I've modified it a slightly.

**Comment by smasonuk** [comment]
This reproduces the error messages with the following steps:

1) Open sln file (vs2022)
2) Set to debug & any cpu
3) Run with debug
4) Wait for molecule to show up (takes a few seconds)
5) Wait 4 more seconds, and close window (close the UI window not the console)

Sometimes (maybe every 1 in 3 times) it will output the message "Released too many references to asset!, or  "[SK error] Unimplemented asset type!"


[TestBug2.zip](https://github.com/StereoKit/StereoKit/files/9462863/TestBug2.zip)

**Reply by maluoi** [reply]
Perfect, thank you! It may be a little later in the week, but I'll dig into this soon and see what I can do to resolve it.

**Reply by smasonuk** [reply]
FYI, I did some digging around,, I found if I commented out these lines I no longer able to get the error message... (I don't actually need these lines anyway). However these lines do not cause an issue on their own.

            var fnt = Font.Default; //Font.FromFile(@"c:\work\Roboto\Roboto-Regular.ttf") ?? 
            _errorTextStyl = Text.MakeStyle(fnt, 2f * U.cm, new Color(1f, 0f, 0f));
            _activeTextStyl = Text.MakeStyle(fnt, 2f * U.cm, new Color(0f, 1f, 0f, 0.7f));
            _activeTextStyl.Material.Transparency = Transparency.Blend;
            _normalTextStyle = Text.MakeStyle(fnt, 2f * U.cm, new Color(1f, 1f, 1f, 0.5f));
            _normalTextStyle.Material.Transparency = Transparency.Blend;

**Reply by smasonuk** [reply]
Also, I build a debug version of SK and this is stack at the point of the error. It's coming from material_release

![image](https://user-images.githubusercontent.com/118555/188288154-4cf39a35-b096-4673-a5eb-a255f95f9c8f.png)


![image](https://user-images.githubusercontent.com/118555/188288192-0bc28f94-9a56-441d-8656-7cf91079e8c3.png)

**Reply by smasonuk** [reply]
I had a further dig into the SK C code.. My C programming skills are severely atrophied, so what i'm about to say may be nonsense!

But could the issue be down to this line in Text.cs:
`public Material Material => new Material(NativeAPI.text_style_get_material(this));`

This create a new Material c# object using the IntPtr, however I can't see anywhere in the code path where it increments the ref counter on the material, However in the Material finalizer it will decrement the ref counter: Causing this issue?
`    NativeAPI.assets_releaseref_threadsafe(_inst);`
